File: packages/python-sdk/src/blindfold/regex/registry.py
"""Detector registry: collects all detectors and filters by locale"""

from typing import Dict, List, Optional, Type

from .base import Detector, RegexDetector, RegionDetector

# Locale to entity type mappings
UNIVERSAL_DETECTORS: List[Type[Detector]] = []
REGION_DETECTORS: Dict[str, List[Type[Detector]]] = {
    "us": [],
    "eu": [],
    "uk": [],
}


def register_universal(cls: Type[Detector]) -> Type[Detector]:
    """Class decorator to register a universal (locale-independent) detector."""
    UNIVERSAL_DETECTORS.append(cls)
    return cls


def register_region(locale: str):
    """Class decorator factory to register a region-specific detector."""
    def decorator(cls: Type[Detector]) -> Type[Detector]:
        if locale not in REGION_DETECTORS:
            REGION_DETECTORS[locale] = []
        REGION_DETECTORS[locale].append(cls)
        return cls
    return decorator


class DetectorRegistry:
    """Builds a list of detector instances filtered by locale."""

    def __init__(
        self,
        locales: Optional[List[str]] = None,
    ) -> None:
        self.locales = [loc.lower() for loc in (locales or ["us"])]
        self._detectors: List[Detector] = []
        self._build()

    def _build(self) -> None:
        for cls in UNIVERSAL_DETECTORS:
            self._detectors.append(cls())

        for locale in self.locales:
            for cls in REGION_DETECTORS.get(locale, []):
                det = cls() if not issubclass(cls, RegionDetector) else cls(locale=locale)
                self._detectors.append(det)

    @property
    def detectors(self) -> List[Detector]:
        return self._detectors

File: packages/js-sdk/src/regex/registry.ts
// Detector registry: collects all detectors and filters by locale

import { Detector } from './base'

// Type for a detector constructor (zero-arg)
type DetectorClass = new () => Detector

// Module-level arrays populated by registerUniversal / registerRegion calls
const UNIVERSAL_DETECTORS: DetectorClass[] = []
const REGION_DETECTORS: Record<string, DetectorClass[]> = {
  us: [],
  eu: [],
  uk: [],
}

/** Register a universal (locale-independent) detector class. */
export function registerUniversal(cls: DetectorClass): void {
  UNIVERSAL_DETECTORS.push(cls)
}

/** Register a region-specific detector class. */
export function registerRegion(locale: string, cls: DetectorClass): void {
  if (!REGION_DETECTORS[locale]) {
    REGION_DETECTORS[locale] = []
  }
  REGION_DETECTORS[locale].push(cls)
}

/** Builds a list of detector instances filtered by locale. */
export class DetectorRegistry {
  private _detectors: Detector[] = []

  constructor(locales?: string[]) {
    const resolvedLocales = (locales ?? ['us']).map((loc) => loc.toLowerCase())
    this._build(resolvedLocales)
  }

  private _build(locales: string[]): void {
    for (const Cls of UNIVERSAL_DETECTORS) {
      this._detectors.push(new Cls())
    }

    for (const locale of locales) {
      const regionClasses = REGION_DETECTORS[locale] ?? []
      for (const Cls of regionClasses) {
        this._detectors.push(new Cls())
      }
    }
  }

  get detectors(): Detector[] {
    return this._detectors
  }
}
